{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://127.0.0.1:8051/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x2496a4c19e8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2024-06-18 15:31:25,413] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\dash.py\", line 1316, in dispatch\n",
      "    callback_context=g,\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\contextvars\\__init__.py\", line 38, in run\n",
      "    return callable(*args, **kwargs)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-11-c32752555b30>\", line 137, in update_styles\n",
      "    } for i in selected_columns]\n",
      "TypeError: 'NoneType' object is not iterable\n",
      "[2024-06-18 15:31:26,802] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\dash.py\", line 1316, in dispatch\n",
      "    callback_context=g,\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\contextvars\\__init__.py\", line 38, in run\n",
      "    return callable(*args, **kwargs)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-11-c32752555b30>\", line 137, in update_styles\n",
      "    } for i in selected_columns]\n",
      "TypeError: 'NoneType' object is not iterable\n",
      "[2024-06-18 15:32:10,436] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\dash.py\", line 1316, in dispatch\n",
      "    callback_context=g,\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\contextvars\\__init__.py\", line 38, in run\n",
      "    return callable(*args, **kwargs)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-11-c32752555b30>\", line 137, in update_styles\n",
      "    } for i in selected_columns]\n",
      "TypeError: 'NoneType' object is not iterable\n",
      "[2024-06-18 15:32:46,909] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py\", line 2897, in get_loc\n",
      "    return self._engine.get_loc(key)\n",
      "  File \"pandas\\_libs\\index.pyx\", line 107, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\index.pyx\", line 131, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "KeyError: 'age_upon_outcome_years'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\dash.py\", line 1316, in dispatch\n",
      "    callback_context=g,\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\contextvars\\__init__.py\", line 38, in run\n",
      "    return callable(*args, **kwargs)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-11-c32752555b30>\", line 95, in update_dashboard\n",
      "    filtered_data = df[df['age_upon_outcome_years'] <= 1].to_dict('records')\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\frame.py\", line 2980, in __getitem__\n",
      "    indexer = self.columns.get_loc(key)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py\", line 2899, in get_loc\n",
      "    return self._engine.get_loc(self._maybe_cast_indexer(key))\n",
      "  File \"pandas\\_libs\\index.pyx\", line 107, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\index.pyx\", line 131, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "KeyError: 'age_upon_outcome_years'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2024-06-18 15:33:03,156] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py\", line 2897, in get_loc\n",
      "    return self._engine.get_loc(key)\n",
      "  File \"pandas\\_libs\\index.pyx\", line 107, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\index.pyx\", line 131, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "KeyError: 'age_upon_outcome_years'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\dash.py\", line 1316, in dispatch\n",
      "    callback_context=g,\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\contextvars\\__init__.py\", line 38, in run\n",
      "    return callable(*args, **kwargs)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-11-c32752555b30>\", line 95, in update_dashboard\n",
      "    filtered_data = df[df['age_upon_outcome_years'] <= 1].to_dict('records')\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\frame.py\", line 2980, in __getitem__\n",
      "    indexer = self.columns.get_loc(key)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py\", line 2899, in get_loc\n",
      "    return self._engine.get_loc(self._maybe_cast_indexer(key))\n",
      "  File \"pandas\\_libs\\index.pyx\", line 107, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\index.pyx\", line 131, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "KeyError: 'age_upon_outcome_years'\n",
      "[2024-06-18 15:33:09,809] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\dash.py\", line 1316, in dispatch\n",
      "    callback_context=g,\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\contextvars\\__init__.py\", line 38, in run\n",
      "    return callable(*args, **kwargs)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-11-c32752555b30>\", line 137, in update_styles\n",
      "    } for i in selected_columns]\n",
      "TypeError: 'NoneType' object is not iterable\n",
      "[2024-06-18 15:33:11,811] ERROR in app: Exception on /_dash-update-component [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py\", line 2897, in get_loc\n",
      "    return self._engine.get_loc(key)\n",
      "  File \"pandas\\_libs\\index.pyx\", line 107, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\index.pyx\", line 131, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "KeyError: 'age_upon_outcome_years'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\flask\\app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\dash.py\", line 1316, in dispatch\n",
      "    callback_context=g,\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\contextvars\\__init__.py\", line 38, in run\n",
      "    return callable(*args, **kwargs)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\dash\\_callback.py\", line 442, in add_context\n",
      "    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%\n",
      "  File \"<ipython-input-11-c32752555b30>\", line 95, in update_dashboard\n",
      "    filtered_data = df[df['age_upon_outcome_years'] <= 1].to_dict('records')\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\frame.py\", line 2980, in __getitem__\n",
      "    indexer = self.columns.get_loc(key)\n",
      "  File \"c:\\users\\jon wickerd\\documents\\snhu\\cs340\\cs-340_local-env\\.venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py\", line 2899, in get_loc\n",
      "    return self._engine.get_loc(self._maybe_cast_indexer(key))\n",
      "  File \"pandas\\_libs\\index.pyx\", line 107, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\index.pyx\", line 131, in pandas._libs.index.IndexEngine.get_loc\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "  File \"pandas\\_libs\\hashtable_class_helper.pxi\", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item\n",
      "KeyError: 'age_upon_outcome_years'\n"
     ]
    }
   ],
   "source": [
    "import dash\n",
    "from dash import dcc, html, dash_table\n",
    "from dash.dependencies import Input, Output\n",
    "import plotly.express as px\n",
    "import dash_leaflet as dl\n",
    "import pandas as pd\n",
    "import base64\n",
    "from crud import AnimalShelter\n",
    "\n",
    "###########################\n",
    "# Data Manipulation / Model\n",
    "###########################\n",
    "\n",
    "\n",
    "# Connect to database via CRUD Module\n",
    "db = AnimalShelter()\n",
    "df = pd.DataFrame.from_records(db.read({}))\n",
    "df.drop(columns=['_id'], inplace=True)\n",
    "\n",
    "# class read method must support return of list object and accept projection json input\n",
    "# sending the read method an empty document requests all documents be returned\n",
    "df = pd.DataFrame.from_records(db.read({}))\n",
    "\n",
    "df.drop(columns=['_id'],inplace=True)\n",
    "\n",
    "## Debug\n",
    "# print(len(df.to_dict(orient='records')))\n",
    "# print(df.columns)\n",
    "\n",
    "\n",
    "#########################\n",
    "# Dashboard Layout / View\n",
    "#########################\n",
    "app = dash.Dash(__name__)\n",
    "\n",
    "\n",
    "image_filename = 'Grazioso Salvare Logo.png' \n",
    "encoded_image = base64.b64encode(open(image_filename, 'rb').read())\n",
    "\n",
    "html.Img(src='data:image/png;base64,{}'.format(encoded_image.decode()))\n",
    "\n",
    "app.layout = html.Div([\n",
    "    html.Center(html.B(html.H1('CS-340 Dashboard Jon Wickerd'))),\n",
    "    html.Img(src='data:image/png;base64,{}'.format(encoded_image), style={'height':'10%', 'width':'10%'}),\n",
    "    html.Hr(),\n",
    "    html.Div(\n",
    "        dcc.Dropdown(\n",
    "            id='filter-type',\n",
    "            options=[\n",
    "                {'label': 'Option 1: Cats', 'value': 'cat'},\n",
    "                {'label': 'Option 2: 1 Year Old', 'value': '1_year'},\n",
    "                {'label': 'Option 3: Pit Bull', 'value': 'pitbull'}\n",
    "            ],\n",
    "            value='cat'  # Default value\n",
    "        )\n",
    "    ),\n",
    "    html.Hr(),\n",
    "    dash_table.DataTable(\n",
    "        id='datatable-id',\n",
    "        columns=[{\"name\": i, \"id\": i, \"deletable\": False, \"selectable\": True} for i in df.columns],\n",
    "        data=df.to_dict('records'),\n",
    "        page_size=10,  # Example pagination\n",
    "        sort_action='native',\n",
    "        filter_action='native',\n",
    "    ),\n",
    "    html.Br(),\n",
    "    html.Hr(),\n",
    "    html.Div(className='row',\n",
    "             style={'display': 'flex'},\n",
    "             children=[\n",
    "                 html.Div(\n",
    "                     id='graph-id',\n",
    "                     className='col s12 m6',\n",
    "                 ),\n",
    "                 html.Div(\n",
    "                     id='map-id',\n",
    "                     className='col s12 m6',\n",
    "                 )\n",
    "             ])\n",
    "])\n",
    "\n",
    "#############################################\n",
    "# Interaction Between Components / Controller\n",
    "#############################################\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "@app.callback(Output('datatable-id','data'),\n",
    "              [Input('filter-type', 'value')])\n",
    "def update_dashboard(filter_type):\n",
    "    if filter_type == 'cat':\n",
    "        filtered_data = df[df['animal_type'] == 'Cat'].to_dict('records')\n",
    "    elif filter_type == '1_year':\n",
    "        filtered_data = df[df['age_upon_outcome_years'] <= 1].to_dict('records')\n",
    "    elif filter_type == 'pitbull':\n",
    "        filtered_data = df[df['breed'] == 'Pit Bull'].to_dict('records')\n",
    "    else:\n",
    "        filtered_data = df.to_dict('records')  # Default to all data if no match\n",
    "\n",
    "    return filtered_data\n",
    "\n",
    "# Display the breeds of animal based on quantity represented in\n",
    "# the data table\n",
    "@app.callback(\n",
    "    Output('graph-id', \"children\"),\n",
    "    [Input('datatable-id', \"derived_virtual_data\")])\n",
    "def update_graphs(viewData):\n",
    "    if viewData is None:\n",
    "        return []\n",
    "\n",
    "    dff = pd.DataFrame.from_dict(viewData)\n",
    "    \n",
    "    # Count occurrences of each breed\n",
    "    breed_counts = dff['breed'].value_counts()\n",
    "    \n",
    "    # Create pie chart only if there are breeds to show\n",
    "    if not breed_counts.empty:\n",
    "        fig = px.pie(names=breed_counts.index, values=breed_counts.values, title='Preferred Animals')\n",
    "        fig.update_layout(\n",
    "            height=400,  # Set a fixed height to avoid disproportionate scaling\n",
    "            width=500,   # Set a fixed width\n",
    "        )\n",
    "        return dcc.Graph(figure=fig)\n",
    "    else:\n",
    "        return html.Div(\"No data available for pie chart.\")\n",
    "    \n",
    "#This callback will highlight a cell on the data table when the user selects it\n",
    "@app.callback(\n",
    "    Output('datatable-id', 'style_data_conditional'),\n",
    "    [Input('datatable-id', 'selected_columns')]\n",
    ")\n",
    "def update_styles(selected_columns):\n",
    "    return [{\n",
    "        'if': { 'column_id': i },\n",
    "        'background_color': '#D2F3FF'\n",
    "    } for i in selected_columns]\n",
    "\n",
    "\n",
    "# This callback will update the geo-location chart for the selected data entry\n",
    "# derived_virtual_data will be the set of data available from the datatable in the form of \n",
    "# a dictionary.\n",
    "# derived_virtual_selected_rows will be the selected row(s) in the table in the form of\n",
    "# a list. For this application, we are only permitting single row selection so there is only\n",
    "# one value in the list.\n",
    "# The iloc method allows for a row, column notation to pull data from the datatable\n",
    "@app.callback(\n",
    "    Output('map-id', \"children\"),\n",
    "    [Input('datatable-id', \"derived_virtual_data\"),\n",
    "     Input('datatable-id', \"derived_virtual_selected_rows\")])\n",
    "def update_map(viewData, index):\n",
    "    if viewData is None or index is None:\n",
    "        return []\n",
    "\n",
    "    dff = pd.DataFrame.from_dict(viewData)\n",
    "    row = index[0] if index else 0\n",
    "    return [\n",
    "        dl.Map(style={'width': '1000px', 'height': '500px'}, center=[30.75, -97.48], zoom=10, children=[\n",
    "            dl.TileLayer(id=\"base-layer-id\"),\n",
    "            dl.Marker(position=[dff.iloc[row, 13], dff.iloc[row, 14]], children=[\n",
    "                dl.Tooltip(dff.iloc[row, 4]),\n",
    "                dl.Popup([\n",
    "                    html.H1(\"Animal Name\"),\n",
    "                    html.P(dff.iloc[row, 9])\n",
    "                ])\n",
    "            ])\n",
    "        ])\n",
    "    ]\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(port=8051, mode='external')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
